File: flutter_app/test/report_timeline_test.dart
import 'package:test/test.dart';
import 'package:splitflow/domain/models.dart';
import 'package:splitflow/domain/report_timeline.dart';

/// `tests/reportTimeline.test.ts` 的 Dart 版。
void main() {
  Expense expense({
    String id = 'e1',
    String title = '晚餐',
    int amount = 1000,
    int? baseAmount = 1000,
    String currency = 'TWD',
    String? date = '2026-03-01',
    String? time,
    ExpenseCategory category = ExpenseCategory.food,
    ExpensePlace? place,
    DateTime? createdAt,
  }) {
    return Expense(
      id: id,
      title: title,
      amount: amount,
      currency: currency,
      baseAmount: baseAmount,
      paidBy: 'u1',
      splits: {'u1': amount},
      category: category,
      date: date,
      time: time,
      place: place,
      createdAt: createdAt,
    );
  }

  test('一天一段，日期由早到晚', () {
    final days = reportTimeline([
      expense(id: 'e1', date: '2026-03-03'),
      expense(id: 'e2', date: '2026-03-01'),
    ], 'TWD');

    expect(days.map((day) => day.date), ['2026-03-01', '2026-03-03']);
  });

  test('同一天照時間由早到晚排', () {
    final entries = reportTimeline([
      expense(id: 'e1', time: '19:00'),
      expense(id: 'e2', time: '08:30'),
      expense(id: 'e3', time: '12:00'),
    ], 'TWD').single.entries;

    expect(entries.map((entry) => entry.time), ['08:30', '12:00', '19:00']);
  });

  test('沒記時間的排在當天最後，不會被猜到中間去', () {
    final entries = reportTimeline([
      expense(id: 'e1', time: '19:00'),
      expense(id: 'e2'),
      expense(id: 'e3', time: '08:30'),
    ], 'TWD').single.entries;

    expect(entries.map((entry) => entry.time), ['08:30', '19:00', '']);
  });

  test('都沒記時間時照記帳先後排，先記的在前', () {
    final entries = reportTimeline([
      // createdAt 還沒回來 = 剛送出的那一筆。
      expense(id: 'e1', amount: 100, baseAmount: 100),
      expense(
        id: 'e2',
        amount: 200,
        baseAmount: 200,
        createdAt: DateTime.utc(2026, 3, 1, 10),
      ),
      expense(
        id: 'e3',
        amount: 300,
        baseAmount: 300,
        createdAt: DateTime.utc(2026, 3, 1, 9),
      ),
    ], 'TWD').single.entries;

    // 先記的（9 點）在前，後記的（10 點）在後。createdAt 還沒回來的那筆是
    // 「剛送出」，排在最後。
    expect(entries.map((entry) => entry.amount), [300, 200, 100]);
  });

  test('每天有小計，加起來就是那天的花費', () {
    final day = reportTimeline([
      expense(id: 'e1', amount: 1000, baseAmount: 1000),
      expense(id: 'e2', amount: 500, baseAmount: 500),
    ], 'TWD').single;

    expect(day.total, 1500);
  });

  test('缺匯率換算不出來的支出不進時間軸 —— 進去的話每日小計會跟總額對不起來', () {
    final day = reportTimeline([
      expense(id: 'e1', amount: 1000, baseAmount: 1000),
      expense(id: 'e2', amount: 5000, baseAmount: null, currency: 'JPY'),
    ], 'TWD').single;

    expect(day.entries, hasLength(1));
    expect(day.total, 1000);
  });

  test('連日期都沒有的支出放不上時間軸', () {
    final days = reportTimeline([expense(date: null)], 'TWD');
    expect(days, isEmpty);
  });

  test('Day 從任務起始日算起，含頭尾', () {
    final days = reportTimeline([
      expense(id: 'e1', date: '2026-03-01'),
      expense(id: 'e2', date: '2026-03-10'),
    ], 'TWD', '2026-03-01');

    expect(days.map((day) => day.day), [1, 10]);
  });

  test('沒有起始日就用最早的那天當 Day 1', () {
    final days = reportTimeline([
      expense(id: 'e1', date: '2026-03-05'),
      expense(id: 'e2', date: '2026-03-06'),
    ], 'TWD');

    expect(days.map((day) => day.day), [1, 2]);
  });

  test('支出早於起始日時改用那天當原點，不會出現 Day 0 或負數', () {
    // 提前買的機票：記在起始日之前。
    final days = reportTimeline([
      expense(id: 'e1', date: '2026-02-01'),
      expense(id: 'e2', date: '2026-03-01'),
    ], 'TWD', '2026-03-01');

    expect(days.first.day, 1);
    expect(days.every((day) => day.day >= 1), isTrue);
  });

  test('只放公開得起的欄位 —— 沒有名稱、沒有地址、沒有人', () {
    // 這份資料會寫進任何人拿到連結都讀得到的文件裡，欄位跑進來就是外洩。
    // Dart 沒有 Object.keys，改成逐一斷言 ReportEntry 只帶得動這四個值。
    final entry = reportTimeline([
      expense(
        title: '阿明的點心',
        time: '15:00',
        place: const ExpensePlace(
          name: '一蘭',
          address: '福岡市博多區',
          lat: 33.59,
          lng: 130.4,
          placeId: 'p_ichiran',
        ),
      ),
    ], 'TWD').single.entries.single;

    expect(entry.time, '15:00');
    expect(entry.category, ExpenseCategory.food);
    expect(entry.amount, 1000);
    // 地點只留名字，座標與地址都不跟著出去。
    expect(entry.place, '一蘭');
  });
}

File: tests/reportTimeline.test.ts
import { describe, expect, it } from "vitest";
import { reportTimeline } from "@/utils/reportTimeline";
import type { Expense } from "@/types/expense";

/** 模擬 Firestore Timestamp：只要有 toDate 就夠了。 */
function at(day: number, hour = 12): Expense["createdAt"] {
  const value = new Date(2026, 2, day, hour);
  return { toDate: () => value } as Expense["createdAt"];
}

function expense(overrides: Partial<Expense> = {}): Expense {
  return {
    id: "e1",
    title: "晚餐",
    category: "food",
    amount: 10000,
    currency: "TWD",
    rate: 1,
    baseAmount: 10000,
    paidBy: "u1",
    splitMode: "even",
    splits: { u1: 10000 },
    place: null,
    receipt: null,
    note: "",
    date: "2026-03-01",
    time: "",
    createdBy: "u1",
    createdAt: at(1),
    ...overrides
  } as Expense;
}

describe("reportTimeline", () => {
  it("一天一段，日期由早到晚", () => {
    const result = reportTimeline(
      [expense({ date: "2026-03-03" }), expense({ date: "2026-03-01" })],
      "TWD"
    );
    expect(result.map(day => day.date)).toEqual(["2026-03-01", "2026-03-03"]);
  });

  it("同一天照時間由早到晚排", () => {
    const result = reportTimeline(
      [
        expense({ id: "dinner", time: "19:30" }),
        expense({ id: "breakfast", time: "08:00" }),
        expense({ id: "lunch", time: "12:15" })
      ],
      "TWD"
    );
    expect(result[0].entries.map(entry => entry.time)).toEqual(["08:00", "12:15", "19:30"]);
  });

  it("沒記時間的排在當天最後，不會被猜到中間去", () => {
    const result = reportTimeline(
      [expense({ id: "unknown" }), expense({ id: "dinner", time: "19:30" })],
      "TWD"
    );
    expect(result[0].entries.map(entry => entry.time)).toEqual(["19:30", ""]);
  });

  it("都沒記時間時照記帳先後排，先記的在前", () => {
    const result = reportTimeline(
      [expense({ id: "second", createdAt: at(1, 20) }), expense({ id: "first", createdAt: at(1, 9) })],
      "TWD"
    );
    expect(result[0].entries.map(entry => entry.amount)).toHaveLength(2);
    // 兩筆金額一樣分不出來，改用名稱驗順序。
    const named = reportTimeline(
      [
        expense({ createdAt: at(1, 20), title: "宵夜" }),
        expense({ createdAt: at(1, 9), title: "早餐" })
      ],
      "TWD"
    );
    expect(named[0].entries.map(entry => entry.name)).toEqual(["早餐", "宵夜"]);
  });

  // 地點整區列在「去過的地方」了，時間軸再列一次就只是同一份資訊講兩遍。
  it("逐筆放的是支出名稱，不是地點", () => {
    const result = reportTimeline(
      [
        expense({
          title: "海南雞飯",
          place: { name: "天天海南雞飯", address: null, lat: null, lng: null, placeId: null }
        })
      ],
      "TWD"
    );

    expect(result[0].entries[0].name).toBe("海南雞飯");
    expect(result[0].entries[0].place).toBeUndefined();
  });

  it("每天有小計，加起來就是那天的花費", () => {
    const result = reportTimeline(
      [expense({ baseAmount: 12000 }), expense({ baseAmount: 3000 }), expense({ date: "2026-03-02", baseAmount: 500 })],
      "TWD"
    );
    expect(result[0].total).toBe(15000);
    expect(result[1].total).toBe(500);
  });

  it("缺匯率換算不出來的支出不進時間軸 —— 進去的話每日小計會跟總額對不起來", () => {
    const result = reportTimeline(
      [expense({ currency: "THB", baseAmount: null }), expense({ baseAmount: 3000 })],
      "TWD"
    );
    expect(result[0].entries).toHaveLength(1);
    expect(result[0].total).toBe(3000);
  });

  it("連日期都沒有的支出放不上時間軸", () => {
    const result = reportTimeline([expense({ date: null, createdAt: null })], "TWD");
    expect(result).toEqual([]);
  });

  it("Day 從任務起始日算起，含頭尾", () => {
    const result = reportTimeline(
      [expense({ date: "2026-03-01" }), expense({ date: "2026-03-05" })],
      "TWD",
      "2026-03-01"
    );
    expect(result.map(day => day.day)).toEqual([1, 5]);
  });

  it("沒有起始日就用最早的那天當 Day 1", () => {
    const result = reportTimeline(
      [expense({ date: "2026-03-03" }), expense({ date: "2026-03-04" })],
      "TWD"
    );
    expect(result.map(day => day.day)).toEqual([1, 2]);
  });

  it("支出早於起始日時改用那天當原點，不會出現 Day 0 或負數", () => {
    // 提前買的機票之類。
    const result = reportTimeline(
      [expense({ date: "2026-02-20" }), expense({ date: "2026-03-01" })],
      "TWD",
      "2026-03-01"
    );
    expect(result[0].day).toBe(1);
    expect(result[1].day).toBe(10);
  });

  describe("每天的天氣", () => {
    const sunny = { code: 0, high: 30, low: 22, exact: null };
    const stormy = { code: 95, high: 28, low: 21, exact: null };

    it("取當天第一筆有天氣的支出", () => {
      const days = reportTimeline(
        [
          expense({ date: "2026-03-01", time: "09:00", weather: stormy }),
          expense({ date: "2026-03-01", time: "18:00", weather: sunny })
        ],
        "TWD"
      );

      expect(days[0].weather).toEqual(stormy);
    });

    it("前面幾筆沒有天氣就往後找", () => {
      const days = reportTimeline(
        [
          expense({ date: "2026-03-01", time: "09:00" }),
          expense({ date: "2026-03-01", time: "18:00", weather: sunny })
        ],
        "TWD"
      );

      expect(days[0].weather).toEqual(sunny);
    });

    it("整天都沒有就是 null，不是硬湊一個", () => {
      const days = reportTimeline([expense({ date: "2026-03-01" })], "TWD");

      expect(days[0].weather).toBeNull();
    });

    it("每一天各自算，不會沿用前一天的", () => {
      const days = reportTimeline(
        [
          expense({ date: "2026-03-01", weather: stormy }),
          expense({ date: "2026-03-02" })
        ],
        "TWD"
      );

      expect(days[0].weather).toEqual(stormy);
      expect(days[1].weather).toBeNull();
    });
  });

  // 名稱是刻意放進來的（時間軸沒有它就只剩時間跟金額），但除此之外
  // 這份資料會寫進任何人拿到連結都讀得到的文件裡，多一個欄位就是多一次外洩。
  it("只放公開得起的欄位 —— 沒有 uid、沒有誰付的、沒有 id", () => {
    const [entry] = reportTimeline([expense({ title: "阿明的點心", time: "15:00" })], "TWD")[0].entries;
    expect(Object.keys(entry).sort()).toEqual(["amount", "category", "name", "time"]);
  });
});
